RFCs/FS-1079-union-properties-visible.md

```diff
@@ -0,0 +1,129 @@
+
+```fsharp
+type Example =
+    | First
+    | Second of int * string
+    
+    // The following members were already generated, and are now visible to F#:
+    
+    member IsFirst : bool
+    
+    member IsSecond : bool
+
+    // The following members were already generated, and remain hidden from F#:
+    
+    member Tag : int
+    
+    type Tags =
+    
+        val First : int
+        
+        val Second : int
+    
+    static member First : Example
+    
+    static member NewSecond : int * string -> Example
+```
+
+# Drawbacks
+
+These attributes might incite beginners to check a value's case using a series of `if x.IsFoo then ... elif x.IsBar then ...` in situations where pattern matching would be more appropriate.
+
+# Alternatives
+
+An alternate design would be to also expose the `Tag` property and `Tags` nested type (#2 through #4 as listed in [Detailed design](#detailed-design)). They can be useful in some limited situations, but are left out of the main proposal for several reasons:
+
+* They are more rarely useful than `Is*`.
+
+* They would be even more prone to confusing beginners, by inciting them to match against `x.Tag` instead of `x`.
+
+* Exposing `Tag` and `Tags` changes whether reordering the cases of a union type declaration is a breaking change:
+
+    |                                     | Before this change   | After this change   |
+    | :---------------------------------- | :------------------: | :-----------------: |
+    | Binary compatible for F# consumer   | No                   | No                  |
+    | Source compatible for F# consumer   | **Yes**              | **No**              |
+    | Binary compatible for C# consumer   | No                   | No                  |
+    | Source compatible for C# consumer   | No                   | No                  |
+
+# Compatiblity
+
+* Is this a breaking change?
+
+No. It was already illegal to explicitly declare members whose name clashes with these generated members.
+
+It does however change what constitutes a breaking change in F# code (see Drawbacks above).
+
+* What happens when previous versions of the F# compiler encounter this design addition as source code?
+
+It fails to compile any invocations of the generated members.
+
+* What happens when previous versions of the F# compiler encounter this design addition in compiled binaries?
+
+They see a normal discriminated union, and do not allow invoking its generated members.
+
+* If this is a change or extension to FSharp.Core, what happens when previous versions of the F# compiler encounter this construct?
+
+It is not directly an extension to FSharp.Core, but it does apply to union types declared in FSharp.Core: `option`, `Result`, `Choice`. Previous versions of the F# compiler do not allow invoking the generated members on these types.
```

Comment:
`Option<'T>` is easy: `IsSome` and `IsNone` are already visible, so there is no change.

For `ValueOption<'T>`, there are also visible `IsSome` and `IsNone`, in addition to the hidden `IsValueSome` and `IsValueNone`. So there are two possible routes:
* make them visible (which makes them redundant with the existing `IsSome`/`IsNone`)
* keep them hidden (which makes `ValueOption<'T>` unique among all union types for not having `Is{{UnionCaseName}}` properties)

I think I'm in favor of the former, for consistency. Plus it's probably simpler to implement than making an exception.

Reply:
I assume we can't remove the existing `IsSome/None`, but adding `IsValueSome/None` would effectively make them aliases of the former, in which they would then probably be unique in our ecosystem.

It's a tough call. For parity you would like to have them, but different members that do the exact same thing brings confusion, and unless the tooltip would clearly describe that they are synonyms, I'm afraid programmers will have to research what the differences are and you'll end up answering questions similar to whether `String` or `string` has better performance / should be preferred / has different semantics etc.

Of course, the answers will be simple, they're equal, but I think it's better to prevent the questions rising in the first place. 

But then there's the argument of parity, auto generated code, predictability, orthogonality etc. It's a tough call ;).

Reply:
I'm not sure what the exact stance on obsoleting API is, but it could be acceptable to mark the existing ValueOption IsSome/IsNone properties `[<Obsolete>]` to reduce confusion.

Reply:
@Tarmil I think your preference here is spot on: hiding `IsValueSome`/`IsValueNone` for ValueOption. Can you make a note of that in the RFC?